Simplify loading examples and fix typo in docs

cthoyt opened the pull request:
This pull request does the following:

1. It simplifies reusing the 5 example ngrams configurations by putting them directly inside the package and creating a small `SmiZip.load` class method that allows referencing them by name. Updates README with an example of this usage.
   ```python
   from smizip import SmiZip

   zipper = SmiZip.load("rdkit.slow")
   zipped = zipper.zip("c1ccccc1C(=O)Cl") # gives bytes
   ```
3. Fixes a broken example in the README - the example files are not themselves just a list but a nested structure inside which the data live
4. Enable CLI to get example n-gram files by name like in:
   ```shell
   smizip    -i test.smi  -o test.smiz  -n rdkit.slow
   smizip -d -i test.smiz -o test.2.smi -n rdkit.slow
5. Enable CLI to get n-gram files by URL like in:
   ```shell
   smizip    -i test.smi  -o test.smiz  -n https://raw.githubusercontent.com/SoseiHeptares/smizip/fda620d72ed1fabfd7c05668528d94ffde9235b8/example-ngrams/combined.slow.json
   smizip -d -i test.smiz -o test.2.smi -n https://raw.githubusercontent.com/SoseiHeptares/smizip/fda620d72ed1fabfd7c05668528d94ffde9235b8/example-ngrams/combined.slow.json
   ```
5. Adds a unit test to make sure everything works!

cthoyt committed:
Enable loading examples from package

cthoyt committed:
Fix typo in docs and add unit test

cthoyt committed:
Update README.rst

cthoyt committed:
Update README.rst

cthoyt committed:
Update README.rst

baoilleach commented:
Sorry about the typo! Thanks for the suggestions, but this needs to work for people using "pip" which will be the majority. Andrew Dalke pointed me to the way to include resources in a pip package, which you can access programmatically but I don't see an easy way for people to access these files from the commandline.

baoilleach commented:
...or perhaps it does work for pip, maybe I've misunderstood.

cthoyt commented:
By default, data files inside the package are included when configuring the package with pyproject.toml (see https://setuptools.pypa.io/en/latest/userguide/datafiles.html#include-package-data)

However, if someone installs from pip, they will have the same issue that it's not obvious where the files are either way, which is why I added git cloning the repo into the example in case they directly want to load data themselves. It is also possible to extend the example to check for http/https and get it over the web to simplify this a bit

cthoyt committed:
Enable loading by URL or example name

cthoyt commented:
@baoilleach I just added functionality to load with URL or also by example name and updated the README to reflect this. This should make it maximally easy for new users to get running.

cthoyt committed:
Update pyproject.toml

baoilleach commented:
Thanks. Will try to find time to review and merge in the next few days.

cthoyt committed:
Make loading more reusable

cthoyt commented:
@baoilleach I made a few more updates that also make this more extensible from the python REPL

baoilleach commented:
Hi @cthoyt, perhaps hold off on further changes as I'm not sure that I want to go too far in this direction. The thing is that I see smizip as "a library for compressing text/SMILES given a set of ngrams", rather than "a library for compressing SMILES". Let me try to explain where I'm coming from.

I don't regard the ngrams I provide as a part of the library itself. I want the user to understand that they need to provide a list of ngrams, and that they need to use the same list for compression as for decompression. That is, the user is responsible for the list of ngrams they use, not the library. The files I provided are simply examples; they are not intended to cover all use cases (e.g. are probably missing support for most elements) and may in fact change over time (e.g. if I add support for ring renumbering, which I left out for this initial release). I want the user to understand this, and to be able to adjust an existing file or make their own (this needs some documentation). To my mind, what I need to do is simply document the curl command to download the examples, rather than add downloading as part of the library or command-line applications.
